### ISSUE TITLE
Formally support previous versions of Istio in Kiali releases

### ISSUE THREAD

**Issue by nrfox** [kiali/kiali#7932]
### What do you want to improve?

When a new version of Kiali is released, only the latest version of Istio at the time of release is supported. Older versions of Istio, while they may work anyways, are not formally supported. This means when performing a revision based upgrade of Istio, Kiali may not operate correctly during the upgrade.

### What is the new behavior?

Kiali should be backward compatible 2 versions of Istio to match Istio's backward compatibility.

When users want to upgrade from one version of Istio to the next using a revision based strategy, they should first upgrade their Kiali instance to match the desired Istio version, then deploy their second Istio revision. Kiali should work with both the old revision and the new revision so that it can continue to be used both during the upgrade and after.

**Comment by jshaughn** [comment]
This is going to be necessary for us to make the Mesh page revision-based update support usable.  Promoting to sprint backlog.

**Comment by jshaughn** [comment]
Thinking on this, it seems the main change is to just update what we say about compatibility, and the associated compatibility chart, on kiali.io.  In general, Kiali has been backward compatible, but we haven't wanted to do additional testing to prove that things are actually working. So, in a addition to the documentation change we need to establish some sort of sanity checking, via test automation, that Kiali and the Kiali operator are working against the same versions listed as supported by upstream Istio. Typically latest, latest-1, latest-2.

Note that we can say something like, "with the noted exceptions", meaning, if we do encounter a difficult back-compat issue, I think we could document the problem rather than spending a large effort on the support.  This is mainly to formalize what we already believe to be true, but have been reluctant to say, that Kiali is backward compatible, and do some level of testing to back up that claim.

**Comment by jshaughn** [comment]
@ScriptingShrimp , Do you have any ideas on how we could automate testing for older versions of Istio?  Our merge CI runs head against the latest Istio (I think), but before a Kiali release we'd basically need to also run against the latest-1 and probably latest-2.  @jmazzitelli , this would be true for molecule tests as well, I think.

**Comment by jmazzitelli** [comment]
> this would be true for molecule tests as well, I think.

The molecule tests support testing with specific Kiali CR `spec.version` values, so it is possible to run the tests against older versions of Kiali Server. Just need to ensure the CI environment is installed with the Istio version we are expecting to test with.

**Comment by ScriptingShrimp** [comment]
@jshaughn yes, upstream gating (GH actions) using `run-integration-tests.sh` where `-iv|--istio-version` defaults to the latest release. This triggers [setup-kind-in-ci.sh](https://github.com/kiali/kiali/blob/9435eeea3495a4378e6c28896e16a77547b567f5/hack/setup-kind-in-ci.sh#L95) where we already have some hard coded conditions for Kaili v1. This will finally trigger `download-istio.sh` which defaults to latest version if not specified otherwise. 

1. Do we specify Istio version during the release? (GH release pipelines - community)

So cloning this upstream release pipeline is very easy solution, but I'm afraid, it will not be sufficient - because of the Molecule test requirements. 

2. Its possible to address the testing of older Istio versions within Jenkin's upstream-istio-pipeline (this one also runs nightly) where we already can specify Istio version to be used. This literally means no extra work besides adding molecule tests - which we want to do for a quite while (see [OSSM-2047](https://issues.redhat.com/browse/OSSM-2047?filter=-1)). Before release we can manually trigger the upstream-istio-pipeline with required istio versions running on top of OCP.

As a side effect I think it would be nice enhancement to introduce machine readable compatibility matrix in the core repository (the same one as we do have on [kiali.io](https://kiali.io/docs/installation/installation-guide/prerequisites/)) so we can use `jq` or `yq` for automation, i.e.:

```yaml
Istio_Compatiblity: 
  - Istio: 1.24 
    Kiali_Min: 2.0.0 
    Kiali_Max: "" 
    Notes: "" 
```
or other way around
```yaml
Kiali_Compatibilty: 
  - Kiali: 2.0.0
    Istio_Recommended: 1.24
    Istio_Minus1: 1.23 
    Istio_Minus2: 1.22
    Notes: "" 
```

**Comment by jshaughn** [comment]
I have a kiali.io PR with updated Prerequisite info and a simplified compatibility table.  I'll add a few folks for review to see if it's in line with what we want to do here.

**Comment by jshaughn** [comment]
> .. cloning this upstream release pipeline is very easy solution

OK, this is good.  What do you think about maybe running the upstream automation nightly against istio latest-1 and latest-2?  We want to have confidence during the sprint that we are backward compatible to those supported versions.  It does not have to be against openshift, I think Kind is sufficient.

> ... but I'm afraid, it will not be sufficient - because of the Molecule test requirements.

I'm not sure what we want to do about the Molecule tests. @jmazzitelli , any suggestion?

> ... As a side effect I think it would be nice enhancement to introduce machine readable compatibility matrix in the core repository ...

I'm less concerned about downstream because for OpenShift I think Istio users should be using OSSM.  Vanilla Istio on OpenShift is a tiny pie slice.  And for OSSM we don't need  a compatibility matrix because we use fixed, supported release combinations.  As such, I think just testing head on OpenShift, against the latest Istio (which should also be the next OSSMe version), works.

**Comment by jmazzitelli** [comment]
Are we talking about testing on non-OpenShift cluster (like KinD)? We already have molecule tests that run nightly on KinD via [this action](https://github.com/kiali/kiali/blob/v2.3.0/.github/workflows/molecules.yml#L37).

It uses [a hack script](https://github.com/kiali/kiali/blob/v2.3.0/hack/ci-kind-molecule-tests.sh) that builds a KinD cluster and tests the latest version of Kiali with the latest istio (but you can [tell it which version of Istio you want to install via -iv option](https://github.com/kiali/kiali/blob/v2.3.0/hack/ci-kind-molecule-tests.sh#L76) and what the [Kiali CR spec.version should be via the -sv option](https://github.com/kiali/kiali/blob/v2.3.0/hack/ci-kind-molecule-tests.sh#L142)):

So we already know how to do this in some automated fashion.

**Comment by jshaughn** [comment]
> Are we talking about testing on non-OpenShift cluster (like KinD)? We already have molecule tests that run nightly on KinD via [this action](https://github.com/kiali/kiali/blob/v2.3.0/.github/workflows/molecules.yml#L37).

I'd say, yes.  All we are trying to do here is have some level of faith that our current kiali release is compatible with istio latest, -1 and -2 (typically the supported versions of istio at kiali-release-time).  OpenShift is important, but as I say above, I'd argue that it's less important for this upstream support exercise, because if we assume that OSSM is the primary Istio impl for OpenShift, then the support model changes to what we support and test for OSSM.

**Comment by jmazzitelli** [comment]
> So, if people agree with what I'm saying above, I think what we want to do is extend our nightly upstream runs (on Kind) to run Kiali heads against Istio latest, -1 and -2. Does that sound right?

sounds right.

**Comment by jmazzitelli** [comment]
I'm going to run molecule tests using that ci hack script and see how it works with the `-iv` option.

Note: here is how we can determine what the latest release of Istio is, along with the -1 and -2 versions (so this gets us the latest 3 versions of Istio):

```bash
curl -s https://api.github.com/repos/istio/istio/releases | jq -r '.[].tag_name' | grep -Eo '^[0-9]+\.[0-9]+' | sort -uV | tail -n 3
```

Today, this returns:
```
1.22
1.23
1.24
```

**Comment by jmazzitelli** [comment]
Here's how I am updating the github workflow for the molecule testing - now you can tell it to run the molecule tests on the last N minor releases of Istio (the default is 3 - the current release, the -1 minor release and the -2 minor release).

https://github.com/kiali/kiali/pull/8039

I kicked off a test run in my fork and I'm watching the action output right now - it looks like its working.

**Comment by jmazzitelli** [comment]
There isn't much Istio specific testing done in the molecule tests (its mainly exercising the operator, not the server, though the server does get some exercise). But we can do the same thing in our other integration tests as I now do in the molecule test action:


```bash
  NUMBER_ISTIO_VERSIONS_TO_TEST="${{ inputs.last_n_istio_versions }}"
  LATEST_ISTIO_VERSIONS="$(curl -s https://api.github.com/repos/istio/istio/releases | jq -r '.[].tag_name' | sort -rV | awk 
-F. '!seen[$1"."$2]++' | head -n "${NUMBER_ISTIO_VERSIONS_TO_TEST}")"
  echo "Testing with the following versions of Istio:"
  echo "${LATEST_ISTIO_VERSIONS}"
  for iv in ${LATEST_ISTIO_VERSIONS}; do
    echo "Testing with Istio version ${iv}"
    ... run hack scripts telling them to install istio version ${iv} and run the tests ...
    ## Delete KinD cluster so we can create a new one with a new version of Istio in the next loop iteration
    kind delete cluster --name $(kind get clusters 2> /dev/null | head -n1)
  done
```

where `${{ inputs.last_n_istio_versions }}` is defined as in input to the workflow:

```yaml
      last_n_istio_versions:
        description: "The last N minor versions of Istio to test. Default is 3 meaning test with the latest version of Istio, along with the previous two minor version releases."
        required: false
        default: 3
        type: number
```

**Comment by jshaughn** [comment]
> But we can do the same thing in our other integration tests as I now do in the molecule test action:

I approved the molecule test pr, that's good.  But we don't want to default to 3 for our merge-CI, it's too much.  I think just the nightly.

**Comment by jshaughn** [comment]
Great, I will merge the kiali.io updated compatibility doco and chart
